wgpu-hal/src/dynamic/mod.rs

```diff
@@ -0,0 +1,67 @@
+use std::any::Any;
+
+use wgt::WasmNotSendSync;
+
+/// Base trait for all resources, allows downcasting via `std::any::Any`.
+pub trait DynResource: Any + WasmNotSendSync + 'static {
+    fn as_any(&self) -> &dyn Any;
+    fn as_any_mut(&mut self) -> &mut dyn Any;
+}
+
+/// Utility macro for implementing `DynResource` for a list of types.
+macro_rules! impl_dyn_resource {
+    ($($type:ty),*) => {
+        $(
+            impl crate::DynResource for $type {
+                fn as_any(&self) -> &dyn std::any::Any {
+                    self
+                }
+
+                fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+                    self
+                }
+            }
+        )*
+    };
+}
+pub(crate) use impl_dyn_resource;
+
+/// Extension trait for `DynResource` used by by implementations of various dynamic resource traits.
+trait DynResourceExt {
+    fn expect_downcast_ref<T: DynResource>(&self) -> &T;
+    fn expect_downcast_mut<T: DynResource>(&mut self) -> &mut T;
+
+    /// Unboxes a `Box<dyn DynResource>` to a concrete type.
+    ///
+    /// # Safety
+    /// - The `Box<dyn DynResource>` must be the correct concrete type.
```

Comment:
style: newline b/w header and list plz

suggestion: I think it's clearer to simply refer to the argument `self` as such; the description of the trait does a pretty good job of making it clear that it's intended to be implemented for `Box<dyn DynResource>`s.

Combined suggestion hunk:

```suggestion
    /// # Safety
    ///
    /// - `self` must be the correct concrete type.
```

Reply:
Implemented with latest force-push.